### PR TITLE
[pushover] Migration of Pushover OH1 action to OH3 binding

### PR DESCRIPTION
**Pull request by cweitkamp**
- Migration of Pushover action to OH3 binding

Replaces #6798

Signed-off-by: Christoph Weitkamp <github@christophweitkamp.de>

### CODE REVIEW

**Comment by wborn** [comment]
I finally gave it a test and it works! :-)
Is it still WIP because you want to make it possible to construct messages using all the other parameters?

**Comment by cweitkamp** [comment]
@wborn Sounds good. I am using this port in my OH3 setup for several weeks and I did not see any error.

I think the list of available methods is finished. Are there any methods you are missing? If so please tell me and I will add it.

From my POV only the README is missing. I will try to finish it soon.

The loss of the `PushoverBuilder` is a loss for this add-on. But your recent [community post](https://community.openhab.org/t/use-own-java-classes-in-rules/106008/28) about external Java classes lets me think about exposing the new builder implementation and add some new actions to send a message by passing a builder. I will give it a try in the upcoming days. But this should not prevent this PR from finalization.

Another idea I have in my mind is to send the content on an ImageItem (`RawType`) attached to a Pushover message directly by passing it to a new action. But that is not urgent too.

**Comment by wborn** [comment]
It would already be nice to have the functionality of this PR available in OH3! The only additional parameter I use is the URL.

When the builder is moved to a non-internal package I think it can also be used in rule engines which would again provide a nice way to build a message with any parameter.

**Comment by cweitkamp** [comment]
I added a `sendURLMessage(message, @Nullable title, url, @Nullable urlTitle)` action.

**Comment by matulekpl** [comment]
can someone please give me a .jar file for this?:)

**Comment by wborn** [comment]
Jenkins made one for this PR here [org.openhab.binding.pushover-3.0.0-SNAPSHOT.jar](https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/addons/bundles/org.openhab.binding.pushover/3.0.0-SNAPSHOT/org.openhab.binding.pushover-3.0.0-SNAPSHOT.jar).

**Comment by randye007** [comment]
Thanks for the jar file. I've successfully installed it and created a thing which is now online.
When I create the rule, I select the Type of Action (send an HTML message) and the Configuration (Pushover Account Thing).
However, I don't see a field for the actual message to send. Can someone tell me what I'm missing? Thx.

**Comment by matulekpl** [comment]
> Thanks for the jar file. I've successfully installed it and created a thing which is now online.
> When I create the rule, I select the Type of Action (send an HTML message) and the Configuration (Pushover Account Thing).
> However, I don't see a field for the actual message to send due to the action limitations outlined above. Can someone please tell me how I can configure the message content? Thx.

How You installed it? I got it in addons folder but nothing happened...

**Comment by matulekpl** [comment]
> Jenkins made one for this PR here [org.openhab.binding.pushover-3.0.0-SNAPSHOT.jar](https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/addons/bundles/org.openhab.binding.pushover/3.0.0-SNAPSHOT/org.openhab.binding.pushover-3.0.0-SNAPSHOT.jar).

weird thing - I added this file to /etc/openhab/addons/ and nothing happens. 
Even after reboot of whole RPI4 there is no such binding on binding:list

**Comment by randye007** [comment]
> How You installed it?

Just create the thing and you will see the option for Pushover. 
I also don't see the binding through the UI or karaf console. Not sure why. Maybe it's because it's an action.

**Comment by matulekpl** [comment]
> > How You installed it?
> 
> Just create the thing and you will see the option for Pushover.
> I also don't see the binding through the UI or karaf console. Not sure why. Maybe it's because it's an action.

I dont see anything with word "pushover" on creating Thing?

**Comment by wborn** [comment]
It works fine for me with the JAR:

pushover

**Comment by matulekpl** [comment]
> It works fine for me with the JAR:
> 
> pushover

That is weird that it doesn’t work at mine.

To what folder did You put this file?

**Comment by randye007** [comment]
@matulekpl  - on my rPi4 running OH3 M2, it's in the directory /usr/share/openhab/addons. Also, check the permissions on your jar file and make sure to assign openhab:openhab to the owner:group. 
I was able to get everything working. You cannot currently use the UI to enter pushover message contents. You need to use scripts.

**Comment by matulekpl** [comment]
> @matulekpl - on my rPi4 running OH3 M2, it's in the directory /usr/share/openhab/addons. Also, check the permissions on your jar file and make sure to assign openhab:openhab to the owner:group.
> I was able to get everything working. You cannot currently use the UI to enter pushover message contents. You need to use scripts.

Thanks for that. Now it is working.

And now another problem;)
Earlier on OH2 in rules it worked like 

`sendPushoverMessage(pushoverBuilder("test message"))
`
And now??

**Comment by randye007** [comment]
Try this:
val actions = getActions("pushover", "pushover:pushover-account:xxxxxxxxx")
actions.sendMessage("Hello World", "openHAB")

It's based on the example given in the README file. 

sendAttachmentMessage()
sendHtmlMessage()
sendURLMessage()

**Comment by cweitkamp** [comment]
I added a final version for the `README.md` and thus do not consider this as WIP any longer.

I did not yet test exposing the message builder.

Another thing which came to my mind was the general structure of Things. It is possible to define multiple users or groups for broadcasting messages. Maybe a Bridge -> Thing structure has to be considered.

File: bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/connection/PushoverMessageBuilder.java [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+            body.addFieldPart(MESSAGE_KEY_TITLE, new StringContentProvider(title), null);
+        }
+
+        if (device != null) {
+            if (device.length() > MAX_DEVICE_LENGTH) {
+                logger.warn("Skip 'device' as it is longer than {} characters. Got: {}.", MAX_DEVICE_LENGTH, device);
+            } else {
+                // TODO [A-Za-z0-9_-]
+                // TODO Messages may be addressed to multiple specific devices by joining them with a comma (such as
+                // device=iphone,nexus5)
```
Comment (cpmeister) [inline bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/connection/PushoverMessageBuilder.java:189]: Can this TODO be done?

File: bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/factory/PushoverHandlerFactory.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    @Override
+    protected @Nullable ThingHandler createHandler(Thing thing) {
+        final ThingTypeUID thingTypeUID = thing.getThingTypeUID();
+
+        if (PUSHOVER_ACCOUNT.equals(thingTypeUID)) {
+            return new PushoverAccountHandler(thing, httpClient);
+        } else if (PUSHOVER_USER.equals(thingTypeUID)) {
+            return null;
```
Comment (cpmeister) [inline bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/factory/PushoverHandlerFactory.java:64]: A thing with no handler?

File: bundles/org.openhab.binding.pushover/src/main/resources/OH-INF/binding/binding.xml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<binding:binding id="pushover" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns:binding="https://openhab.org/schemas/binding/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/binding/v1.0.0 https://openhab.org/schemas/binding-1.0.0.xsd">
+
+	<name>Pushover Binding</name>
+	<description>Pushover - Simple Notifications.</description>
+	<author>Christoph Weitkamp</author>
```
Comment (cpmeister) [inline bundles/org.openhab.binding.pushover/src/main/resources/OH-INF/binding/binding.xml:8]: The authoring field is deprecated now.

File: bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/connection/PushoverMessageBuilder.java [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+                    throw new IllegalArgumentException(
+                            String.format("Skip sending the message as 'urlTitle' is longer than %d characters.",
+                                    MAX_URL_TITLE_LENGTH));
+                }
+                body.addFieldPart(MESSAGE_KEY_URL_TITLE, new StringContentProvider(urlTitle), null);
+            }
+        }
+
+        if (sound != null) {
+            // TODO validate sound
```
Comment (cpmeister) [inline bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/connection/PushoverMessageBuilder.java:241]: can this be done?

**Review by cpmeister — CHANGES_REQUESTED** [review]
You also need to try and address the build warnings.

File: bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/connection/PushoverMessageBuilder.java (hunk 4 above)
Comment (cweitkamp) [reply bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/connection/PushoverMessageBuilder.java]: Not needed anymore. Sounds are now provided by a `ConfigOptionsProvider`.

File: bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/connection/PushoverMessageBuilder.java (hunk 1 above)
Comment (cweitkamp) [reply bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/connection/PushoverMessageBuilder.java]: Not needed anymore. Can be passed by parameter directly.

File: bundles/org.openhab.binding.pushover/src/test/java/org/openhab/binding/pushover/internal/actions/PushoverActionsTest.java [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+                TITLE, PushoverMessageBuilder.EMERGENCY_PRIORITY));
+    }
+
+    // @Test
+    // public void testSendPriorityMessageWithoutMessage() {
+    // pushoverThingActions.setThingHandler(mockPushoverAccountHandler);
+    // Optional<String> message = Optional.of(null);
+    // assertThrows(IllegalArgumentException.class, () -> PushoverActions.sendPriorityMessage(pushoverThingActions,
+    // message.get(), TITLE, PushoverMessageBuilder.EMERGENCY_PRIORITY));
+    // }
```
Comment (cpmeister) [inline bundles/org.openhab.binding.pushover/src/test/java/org/openhab/binding/pushover/internal/actions/PushoverActionsTest.java:178]: same here

File: bundles/org.openhab.binding.pushover/src/test/java/org/openhab/binding/pushover/internal/actions/PushoverActionsTest.java [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+    }
+
+    // @Test
+    // public void testCancelPriorityMessageWithoutReceipt() {
+    // pushoverThingActions.setThingHandler(mockPushoverAccountHandler);
+    // Optional<String> receipt = Optional.of(null);
+    // assertThrows(IllegalArgumentException.class,
+    // () -> PushoverActions.cancelPriorityMessage(pushoverThingActions, receipt.get()));
+    // }
+
```
Comment (cpmeister) [inline bundles/org.openhab.binding.pushover/src/test/java/org/openhab/binding/pushover/internal/actions/PushoverActionsTest.java:215]: and here

File: bundles/org.openhab.binding.pushover/src/test/java/org/openhab/binding/pushover/internal/actions/PushoverActionsTest.java [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+    public void testSendMessageThingHandlerIsNull() {
+        assertThrows(RuntimeException.class, () -> PushoverActions.sendMessage(pushoverThingActions, MESSAGE, TITLE));
+    }
+
+    // @Test
+    // public void testSendMessageWithoutMessage() {
+    // Optional<String> message = Optional.of(null);
+    // assertThrows(IllegalArgumentException.class,
+    // () -> PushoverActions.sendMessage(pushoverThingActions, message.get(), TITLE));
+    // }
```
Comment (cpmeister) [inline bundles/org.openhab.binding.pushover/src/test/java/org/openhab/binding/pushover/internal/actions/PushoverActionsTest.java:93]: and here

**Review by cpmeister — APPROVED** [review]
Other than these last few comments. LGTM

File: bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/handler/PushoverAccountHandler.java [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+        if (configValid) {
+            updateStatus(ThingStatus.UNKNOWN);
+
+            connection = new PushoverAPIConnection(httpClient, config);
+            try {
+                connection.validateUser();
+                updateStatus(ThingStatus.ONLINE);
+            } catch (PushoverCommunicationException | PushoverConfigurationException e) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getMessage());
+            }
```
Comment (cpmeister) [inline bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/handler/PushoverAccountHandler.java:93]: `connection.validateUser` might be slow, so please make this code asynchronous so the `initialize()` method remains speedy.

**Review by cpmeister — CHANGES_REQUESTED** [review]
Spoke too soon. Final comment.

File: bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/handler/PushoverAccountHandler.java (hunk 8 above)
Comment (cweitkamp) [reply bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/handler/PushoverAccountHandler.java]: Good catch. I somehow had in my mind I planned to do that.

**Review by cpmeister — APPROVED** [review]

**Review by wborn — APPROVED** [review]
Thanks a lot for the migration :+1: The code looks great too so let's merge it! :-)

**Comment by kaikreuzer** [comment]
@cweitkamp This part of the README looks a bit broken: https://next.openhab.org/addons/bindings/pushover/#thing-actions

**Comment by cweitkamp** [comment]
Yes, looks weird. I submitted https://github.com/openhab/openhab-addons/pull/9184 to fix it.

**Comment by fxpde** [comment]
@cweitkamp, thanks for the migration. Works fine with the latest snapshot for me.
Just one proposal to make it easier to use with the new UI. Could you add the binding to the set of actions available when creating a rule? habot for example is listed there.
image

**Comment by cweitkamp** [comment]
We hide all binding actions in UIs until a solution for https://github.com/openhab/openhab-core/issues/1745 has been implemented.

**Pull request merged**